rfd/0154-logging-guidelines.md

```diff
@@ -0,0 +1,280 @@
+>"device_id": ref.DeviceID,
+>"os_type":   ref.OSType,
+>"asset_tag": assetTag,
+>})
+>...
+>logrusLogger.Debug("some message")
+>...
+>logrusLogger.Warn("some other message")
+>```
+>`log/slog`
+>```go
+>slogLogger := slog.With("device_id", ref.DeviceID, "os_type", ref.OSType, "asset_tag", assetTag)
+>...
+>slogLogger.DebugContext(ctx, "some message")
+>...
+>slogLogger.WarnContext(ctx, "some other message")
+>```
+
+#### WithError
+
+Unlike `logrus`, `log/slog` does not provide any special APIs for dealing with errors. When converting to `log/slog` any
+usages of `WithError` should follow the same advice from the `WithField(s)` section. If the error is only included in a
+single log message, then it should be added as an attribute to the log function. If the error is included in multiple
+messages then a logger should be created via `With("error", err)`.
+
+> `logrus`
+>```go
+>logrus.WithError(err).Error("Error parsing response.")
+>```
+>`log/slog`
+>```go
+>slog.ErrorContext(ctx, "Error parsing response.", "error", err)
+>```
+
+### Best Practices
+
+If many log lines share a common attribute, use a `slog.With` to construct a `slog.Logger` with that attribute. This
+allows it to only be formatted once instead of every time a log is emitted.
+
+If an attribute is only emitted in a single log entry, prefer passing it as an attribute instead of
+calling `slog.With`.
+
+Arguments to a logging function are always evaluated, even if the log event is discarded due to configured verbosity.
+If the object implements `fmt.Stringer` you can pass it by pointer and the `String` method will only be called when
+the log is being written. Another option is to implement
+the [`slog.LogValuer`](https://pkg.go.dev/log/slog@master#LogValuer) interface.
+
+For high volume log messages prefer to use `slog.Logger.LogAttrs`. While the API is more verbose, it avoids any
+allocations, making it the most efficient way to produce output. The example below illustrates how to convert a
+message using `InfoContext` to `LogAttrs`.
+
+```go
+slog.InfoContext(ctx, "Speed threshold reached", "flux_capacitors_charged", true, "operator", "doc brown")
+slog.LogAttrs(ctx, slog.LevelInfo, "Speed threshold reached", slog.Bool("flux_capacitors_charged", true), slog.String("operator", "doc brown"))
+```
+
+### Logging Standards
+
+There currently exists no logging standards within Teleport, which has resulted in various different patterns depending
+on which file you open. If we are going to take the time to adjust all existing log messages while converting
+to `log/slog`, we should do our best to start following a few simple rules.
+
+1) Embrace structured logging. Prefer using static messages and attributes instead of relying on `fmt.Sprintf` style
+   formatting.
+1) All keys must be snake case to ensure all output from Teleport is consistent.
+1) Use the Context variants of the `slog.Logger` API. Providing the `context.Context` to all log messages allows them
```

Comment:
One final confirmation: are methods without a context supposed to use context.Background()?

Reply:
I'd echo the preference towards `context.TODO()` here if no context is available.

Reply:
I would discourage `context.TODO()` if there's no intent to follow up, which I would guess applies to a good chunk of methods without a ctx.